Global styles: convert preset font size values to CSS vars

ramonjd opened the pull request:
Parent issue: https://github.com/WordPress/gutenberg/issues/44888

Context: https://github.com/WordPress/gutenberg/issues/44752#issuecomment-1276907842

## What?
This PR converts font size presets to `var:preset|font-size|${slug}` values so that they can be parsed by `getValueFromVariable` and converted to `var(--wp--preset--font-size--${slug}`) values.

## Why?

Why not?

At the moment we're saving the raw value. When fluid is turned on this means we're saving the `clamp()` value.

CSS vars are there. Why not use them?

## Testing

In the site editor, apply a font size preset to an element/block in global styles (not block styles)

<img width="1007" alt="Screen Shot 2022-10-14 at 6 03 08 pm" src="https://user-images.githubusercontent.com/6458278/195783498-430a4f0b-7bac-4c2b-8cac-0d20e72b7a86.png">

Check that we're using the preset CSS variable, and not the raw font size value.

ramonjd committed:
messing around seeing how we can use the preset slug + css var instead of the clamp value in global styles.

ramonjd committed:
Fixed import and formatting

ramonjd committed:
Create a new hook specifically for font size.

ramonjd commented on packages/components/src/font-size-picker/index.tsx:
```diff
@@ -230,7 +230,11 @@ const UnforwardedFontSizePicker = (
 											? selectedItem.size
-											: Number( selectedItem.size )
+											: Number( selectedItem.size ),
+										getSelectedOption(
```
Would there be a better way to get the preset slug back to the callback?

The reason we need it is so we can build the `var:preset|font-size|${slug}` value.

andrewserong replied:
I quite like this approach in that you're passing the whole selected preset in the `onChange` callback. That means that the consumer can potentially do other things with that metadata, rather than passing only the slug.

andrewserong commented on packages/edit-site/src/components/global-styles/typography-panel.js:
```diff
@@ -110,6 +110,25 @@ function useStyleWithReset( path, blockName ) {
+		}
+		setStyleCallback( newValue );
+	};
+	return [ style, setStyle, hasStyle, resetStyle ];
```
This is a bit of a nit: Just double-checking, but is the purpose of using an array here so that it can be renamed? In the case of `useStyleWithReset` that appears to be the case as it's used many times in the TypographyPanel, but `useFontSizeStyleWithReset` is only used once and is particular to `fontSize` rather than a generic hook to be reused in different cases. So, in this case, would an object return type be slightly safer?

This is definitely not a blocker as this hook isn't exported anywhere, I'm just usually slightly wary of arrays as return types because it's easy to accidentally make a mistake with the positioning in later usage.

ramonjd replied:
Yeah, it was a copy and paste job. Thanks for feedback, definitely agree that we can change it to an object.

Cheers!

andrewserong commented on packages/edit-site/src/components/global-styles/typography-panel.js:
```diff
@@ -110,6 +110,25 @@ function useStyleWithReset( path, blockName ) {
+	const hasStyle = () => !! userStyle;
+	const resetStyle = () => setStyleCallback( undefined );
+	const setStyle = ( newValue, metadata ) => {
+		// Convert font size styles to fluid if fluid is activated.
+		if ( isFluidTypographyEnabled && !! metadata?.slug ) {
+			newValue = `var:preset|font-size|${ metadata?.slug }`;
```
Why are we only storing using the `var:` line if fluid typography is enabled? I was wondering if it would also be useful to store the `var:` form for non-fluid typography theme so that we still store a reference to the `var()` rather than the explicit CSS value stored in the preset 🤔

ramonjd replied:
That is a good question. I had a problem and I was working backwards 😄 
Makes total sense to use the CSS var whenever a preset is available. I'll update.

andrewserong reviewed (commented):
This is testing pretty well for me @ramonjd! Just left a couple of comments, one nit, and one question about whether we should also allow this behaviour on themes that haven't opted-in to fluid typography.

Also, since this updates the font size picker in the components package, shall we add a changelog entry, too?

I'll just CC @ciampo and @mirka in case they have input about the update to the signature of the `onChange` callback, too 🙂

Thanks for the follow-up, it'll be great to store the `var:` form of fluid values in particular, rather than the `clamp()` value, and it'll be really good to have this in there for when we eventually allow custom control over the fluid typography rules, too 👍

ciampo commented:
> I'll just CC @ciampo and @mirka in case they have input about the update to the signature of the `onChange` callback, too 🙂

Thank you for the ping! Regarding the `onChange` signature change, I don't see major issues — at most, just a matter of consistency in the way our `onChange` callbacks work across the package. @mirka , any opinions here?

> Also, since this updates the font size picker in the components package, shall we add a changelog entry, too?

Yes!

On top of that, I think it would be great if we could coordinate with @noisysocks and [his PR adding more better unit tests](https://github.com/WordPress/gutenberg/pull/45298) so that we can add good test coverage for this change too.

Another preference is to try and keep changes to `FontSizeKeeper` separate from the changes to the `edit-site` package — but I wouldn't necessarily call it a blocker in this case, especially given the overall size of the PR.

ramonjd committed:
Refactor useFontSizeStyleWithReset so that it exports named methods and values.
Ensure that we allow presets for all font size preset, not only while fluid typography is activated

ramonjd committed:
CHANGELOG.md update

ramonjd commented:
Thanks for the reminder about the changelog, folks!

I had intended to get unit tests done, but might wait until https://github.com/WordPress/gutenberg/pull/45298 is merged until I do.

🙇

noisysocks commented:
https://github.com/WordPress/gutenberg/pull/45298 is merged! The test coverage for `FontSizePicker` is pretty good now so let's keep it that way.

ramonjd committed:
Updated onChange mock arg expectations to include the second arg: fontSize object argument.

ramonjd commented on packages/components/src/font-size-picker/test/index.tsx:
```diff
@@ -395,7 +410,12 @@ describe( 'FontSizePicker', () => {
-				expect( onChange ).toHaveBeenCalledWith( expectedValue );
+				expect( onChange ).toHaveBeenCalledWith(
+					expectedValue,
+					fontSizes.find(
```
@noisysocks Did I just butcher your tests? 😄 

Would you think it's better to be explicit about expected mock args here like I've done [above](https://github.com/WordPress/gutenberg/pull/44967/files#diff-e9acdac412fe0b8c73464e6678c0c58d73f01a10d175ff0ec33bbb4bef3ba9c2R199)?

`expect( onChange.mock.calls ).toEqual( [ expectedValue ]`

noisysocks replied:
I personally prefer an explicit assertion so that the tests don't become an exact replica of the code being tested. 

You might also use `toHaveBeenCalledWith( ...expectedArguments )` instead of `mock.calls` because reads a little more like English. 

Doesn't really matter though. Developer preference!

ramonjd replied:
Good call, I like! Will update, thanks.

ramonjd committed:
Updating tests to include explicit args for mock onChange expectations

ciampo reviewed (commented):
`FontSizePicker`'s code changes are looking good! I'll let @noisysocks give the final blessing ✨

ramonjd committed:
Use find as an alternative to the deleted `getSelectedFontSize` method to pass the immediate, currently-selected font size object containing preset slug information to the consuming parent component.
Updated tests.
General post-rebase chicanery

ramonjd committed:
Now that we're setting font sizes using their presets we don't have to clampify incoming values for the site editor font size picker

ramonjd commented:
@noisysocks Could you give this one a sanity check?

Also pass your eye on the fix for the issue highlighted in https://github.com/WordPress/gutenberg/pull/44966#issuecomment-1301553622

Thank you, sir!

ramonjd committed:
Remove confusing comment, because it had nothing to do with the code changes.

noisysocks commented on packages/edit-site/src/components/global-styles/typography-panel.js:
```diff
@@ -110,6 +109,26 @@ function useStyleWithReset( path, blockName ) {
 	return [ style, setStyle, hasStyle, resetStyle ];
 }
 
+function useFontSizeStyleWithReset( path, blockName ) {
```
nit: `useFontSizeWithReset` or `useFontSize` would be more consistent with the other hooks in this file.

ramonjd replied:
Good point. On it.

noisysocks reviewed (approved):
The `FontSizePicker` code changes look good. I gave it a spin and see `var()` values that point to `clamp()` values when selecting a preset and `clamp()` values when inputting a custom font size.

👍

andrewserong reviewed (approved):
Nice one, this is testing well for me @ramonjd!

✅ Tested that setting existing preset values for Typography in global styles prior to applying this PR results in those preset values still showing up as selected after applying this PR
✅ Tested that with this PR applied, setting Typography based on presets works as expected, the CSS var is applied rather than the raw clampified value, and after a reload the selected values display as they should.

LGTM, and I like the suggestion from @noisysocks to potentially rename the hook, too 👍

ramonjd committed:
Updating method name for clarity.
